bin/helpers/utils.js

```diff
@@ -311,25 +325,187 @@ exports.getLocalFlag = (connectionSettings) => {
+    let local_mode = 'on-demand';
+    if (!this.isUndefined(args.localMode) && args.localMode == 'always-on') {
+      local_mode = 'always-on';
+    } else if (
+      !this.isUndefined(bsConfig['connection_settings']['local_mode']) &&
+      bsConfig['connection_settings']['local_mode'].toLowerCase() ===
+        'always-on'
+    ) {
+      local_mode = 'always-on';
+    }
+    bsConfig['connection_settings']['local_mode'] = local_mode;
+    if (this.isUndefined(args.sync) || !args.sync ){
+      bsConfig['connection_settings']['sync_inferred'] = true;
+    }
+    args.sync = true;
+
+    let localModeInferred = !(this.searchForOption('--local-mode'));
+
+    if (localModeInferred) {
+      bsConfig.connection_settings.local_mode_inferred = local_mode;
+    }
+  }
+};
+
+exports.setupLocalTesting = (bsConfig, args) => {
+  return new Promise(async (resolve, reject) => {
+    let localIdentifierRunning = await this.checkLocalIdentifierRunning(
+      bsConfig, bsConfig['connection_settings']['local_identifier']
+    );
+    if (bsConfig['connection_settings']['local'] && !localIdentifierRunning){
+      var bs_local = new browserstack.Local();
+      var bs_local_args = this.setLocalArgs(bsConfig, args);
+      let that = this;
+      logger.info('Setting up Local testing...');
+      bs_local.start(bs_local_args, function (localStartError) {
+        if (that.isUndefined(localStartError)) {
+          resolve(bs_local);
+        } else {
+          let message = `name: ${localStartError.name}, message: ${localStartError.message}, extra: ${localStartError.extra}`,
+              errorCode = "local_start_error";
+          that.sendUsageReport(
+            bsConfig,
+            args,
+            message,
+            Constants.messageTypes.ERROR,
+            errorCode
+          );
+          reject(localStartError);
+        }
+      });
+    } else {
+      resolve();
+    }
+  });
+};
+
+exports.stopLocalBinary = (bsConfig, bs_local) => {
+  return new Promise((resolve, reject) => {
+    if (!this.isUndefined(bs_local) && bs_local.isRunning() && bsConfig['connection_settings']['local_mode'].toLowerCase() != "always-on") {
+      let that = this;
+      bs_local.stop(function (localStopError) {
+        if (that.isUndefined(localStopError)) {
+          resolve();
+        } else {
+          let message = `name: ${localStartError.name}, message: ${localStartError.message}, extra: ${localStartError.extra}`,
+              errorCode = "local_stop_error";
+          that.sendUsageReport(
+            bsConfig,
+            args,
+            message,
+            Constants.messageTypes.ERROR,
+            errorCode
+          );
+        }
+      });
+    } else {
+      resolve();
+    }
+  });
+};
+
+exports.setLocalArgs = (bsConfig, args) => {
+  let local_args = {}
+  local_args['key'] = bsConfig['auth']['access_key'];
+  local_args['localIdentifier'] = bsConfig["connection_settings"]["local_identifier"];
```

Comment:
We should check the availability of `bsConfig['connection_settings']`

Reply:
We already have a check-in setupLocalTesting.